Cloudflare ad network example

oliy opened the pull request:
For issue [#6881](https://github.com/ampproject/amphtml/issues/6881)

This adds example code for implementing an ad network that uses the Cloudflare A4A signing service.

dknecht commented:
Can we still get them in same release?

keithwrightbos commented:
Very likely, next canary cut should be Wednesday of next week.

On Thu, Jan 5, 2017 at 9:24 AM dknecht <notifications@github.com> wrote:

> Can we still get them in same release?
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/6882#issuecomment-270655245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABr0Lzj0eC252RjT9NGOvNahDr4n_3V8ks5rPP0qgaJpZM4LbPPe>
> .
>

jridgewell commented on ads/_a4a-config.js:
```diff
@@ -44,6 +48,7 @@ import {map} from '../src/utils/object';
 export const a4aRegistry = map({
   'adsense': adsenseIsA4AEnabled,
   'doubleclick': doubleclickIsA4AEnabled,
+  'cloudflare': cloudflareIsA4AEnabled,
```
This should only be enabled in dev mode, right?

dknecht replied:
We will be having some production demos so would like to keep enabled.

oliy commented:
I got the lint and type check errors out of the way, but I see a dist timeout.  Is that an intermittent error?

keithwrightbos commented:
I noticed that #6884 was closed.  Are you planning to use this PR to include the triplelift A4A implementation?  If so, can you address my review comments in the original PR?

oliy commented:
I think I've addressed all your comments from the triple lift implementation.  This won't include the triple lift implementation, but hopefully, it will inform them of changes necessary to have a clean submission.  I'll be adding in tests this coming week.

keithwrightbos commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,109 @@
+    //   const rect = this.getIntersectionElementLayoutBox();
+    // can extract the layout size and be used to pass layout hints to the server.
+
+    // Be sure that this doesn't allow the generation of ANY url, since it could
```
How/when do you plan to add checks?

keithwrightbos commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,109 @@
+   * Extract creative and signature from a Cloudflare signed response.
+   *
+   * Note: Invalid A4A content will NOT have a signature, which will automatically
+   *   cause the A4A processing to fall through to the 3p ad flow.
```
Note that lack of signature does not actually cause the 3p ad flow to be used where a xdomain iframe is created in which custom JS executes to generate an ad request.  Instead, it will take the creative given and render it directly within a cross domain frame using either network cache (if Android) or safeframe (if iOS).

keithwrightbos commented on extensions/amp-ad-network-cloudflare-impl/0.1/cloudflare-a4a-config.js:
```diff
@@ -0,0 +1,29 @@
+  // check for an explicit attribute to enable A4A processing.
+  // This could be just true, if all ads from this network are A4A, or any
+  // other logic based upon configuration, client settings, etc.
+  return !!element.getAttribute('data-a4a');
```
Could add check for src attribute existence

oliy commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,109 @@
+   * @param {!Element} element
+   */
+  /*
+  constructor(element) {
```
I wanted this file to serve as an example starting point for other implementations.  Since many of them actually will need initialization, I've included most of the boilerplate here for guidance/convenience and given instructions to remove this in actual implementations.  If you have a strong preference, I'm fine with removing this.

oliy replied:
Is this blocking the merge?  If so, I will remove this.

lannka replied:
I agree with @keithwrightbos . let's remove. we can make the `fake-impl` as a template.

oliy replied:
Removing.

oliy commented:
I've added tests and made the most recently requested changes.  Let me know if you'd like anything else.  I'd like this to be a clean template for future implementers to start with to integrate Cloudflare ad signing.

dknecht commented:
cc / @keithwrightbos

keithwrightbos commented:
Can you explain why it is necessary to create a completely separate extension from the fake-impl?  Why not just allow for fake-impl to extract its signing services via an element attribute?

tdrl commented:
Sorry I'm late to this conversation.  Just so I'm clear: Is Cloudflare providing an _ad network_ service?  Or only a signing service for other ad networks (e.g., TripleLift)?

tdrl commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,122 @@
+    // be sure that this only generates urls to valid hosts, since it could
+    // allow cookie leaks, etc.
+
+    return CLOUDFLARE_BASE_URL + this.element.getAttribute('src');
```
Since the 'src' attribute is publisher-controlled, is there a possibility for hostile content injection here?  Would it be feasible to uriEncode the src?

oliy replied:
added encoding.

dknecht commented:
@tdrl  @keithwrightbos  We are providing a signing service, but would like to be able to do a full working demo to help make educating potential users easier.  We will use for PSA on our blog's amp content.

keithwrightbos commented:
And you feel that modifying fake as suggested would be insufficient?  I'm concerned about making several extensions that will never actually be used.  @jridgewell your thoughts?

jridgewell commented:
This is more a question for @lannka or @cramforce.

lannka commented on build-system/server.js:
```diff
@@ -19,7 +19,8 @@
  */
 var BBPromise = require('bluebird');
-var app = require('express')();
+var express = require('express');
```
is `express` used?

oliy replied:
I had worked around needing this, so this change is no longer needed.  I'm removing it.

dknecht commented:
I don't believe the fake-impl can be used in production.

cramforce commented:
I think having such demo code makes sense at this stage to help mature the content. Lets proceed with regular review.

dknecht commented:
@keithwrightbos  are we good to go?  We are talking with @jasti about ways to remove in the future.

lannka commented on extensions/amp-ad-network-cloudflare-impl/0.1/test/test-amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,134 @@
+
+import {AmpAdNetworkCloudflareImpl} from '../amp-ad-network-cloudflare-impl';
+import {
+  AmpAdUIHandler, // eslint-disable-line no-unused-vars
```
remove ?

oliy replied:
Done.

lannka commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,110 @@
+    // be sure that this only generates urls to valid hosts, since it could
+    // allow cookie leaks, etc.
+
+    return CLOUDFLARE_BASE_URL + encodeURI(this.element.getAttribute('src'));
```
encodeURIComponent

oliy replied:
Needs to be encodeURI, since this is meant to be the continuation of the URL, to support path references, etc.

lannka commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,110 @@
+ * @type {string}
+ * @private
+ */
+//const CLOUDFLARE_BASE_URL = 'http://www.mycloudflaredomain.com/_a4a';
```
can we remove this and merge JsDoc into the other one?

oliy replied:
Done.

lannka commented on extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js:
```diff
@@ -0,0 +1,110 @@
+ * @private
+ */
+const CLOUDFLARE_BASE_URL =
+  '/extensions/amp-ad-network-cloudflare-impl/0.1/data';
```
we can use simple string like: `/cloudflare/a4a-data`

oliy replied:
This follows the pattern from the fake impl to locate test data, so this represents the path to the test data folder.

lannka commented on build-system/server.js:
```diff
@@ -556,6 +556,51 @@ app.use('/bind/form/get', function(req, res, next) {
+ * Handle CORS headers
+ */
+app.use(cloudflareDir, function fakeCors(req, res, next) {
+    res.setHeader("Access-Control-Allow-Methods", "GET,OPTIONS");
```
can we reuse the existing method `assertCors`?

lannka commented on build-system/server.js:
```diff
@@ -556,6 +556,51 @@ app.use('/bind/form/get', function(req, res, next) {
+/**
+ * Handle fake data
+ */
+app.get(cloudflareDir + '/*', function(req, res) {
```
can we reuse the existing code at line 543?

you do:
```
app.get([doubleclickPath, cloudflarePath], function(req, res) {
}
```

oliy replied:
This one is different, since it has different header handling than the fake one.

lannka replied:
is `X-AmpAdSignature` the only difference ? we can add it to that method.

lannka commented on build-system/server.js:
```diff
@@ -556,6 +556,51 @@ app.use('/bind/form/get', function(req, res, next) {
+/**
+ * Handle CORS preflight
+ */
+app.options(cloudflareDir + '/*', function(req, res) {
```
are we able to merge this into the previous CORS handler?

oliy commented:
The test failure only occurred after rebasing on master.  I'm pretty sure we didn't do anything that would have changed that test result.  Also, it looks like timing could affect the test results.

cramforce commented:
Restarted the test. You can always force-push a new git-sha to get a new run. Apologies for the flaky test.

lannka commented on build-system/server.js:
```diff
@@ -556,6 +556,46 @@ app.use('/bind/form/get', function(req, res, next) {
+  assertCors(req, res, ['GET', 'OPTIONS'], ['X-AmpAdSignature']);
+
+  if (req.method=='OPTIONS') {
+    res.status(204);
```
nit: `res.status(204).end()`

oliy replied:
Done.

oliy commented:
@lannka I think I implemented all changes from your review.
